src/conduits/RwaInputConduit3.sol

```diff
@@ -0,0 +1,297 @@
+    }
+
+    /*//////////////////////////////////
+               Administration
+    //////////////////////////////////*/
+
+    /**
+     * @notice Updates a contract parameter.
+     * @param what The changed parameter name. `"quitTo", "to"`
+     * @param data The new value of the parameter.
+     */
+    function file(bytes32 what, address data) external auth {
+        if (what == "quitTo") {
+            quitTo = data;
+        } else if (what == "to") {
+            to = data;
+        } else {
+            revert("RwaInputConduit3/unrecognised-param");
+        }
+
+        emit File(what, data);
+    }
+
+    /*//////////////////////////////////
+               Operations
+    //////////////////////////////////*/
+
+    /**
+     * @notice Swaps the GEM balance of this contract into DAI through the PSM and push it into the `to` address.
+     * @dev `msg.sender` must have received push access through `mate()`.
+     */
+    function push() external isMate {
+        _doPush(gem.balanceOf(address(this)));
+    }
+
+    /**
+     * @notice Swaps the specified amount of GEM into DAI through the PSM and push it into the `to` address.
+     * @dev `msg.sender` must have received push access through `mate()`.
+     * @param amt Gem amount.
+     */
+    function push(uint256 amt) external isMate {
+        _doPush(amt);
+    }
+
+    /**
+     * @notice Flushes out any GEM balance to `quitTo` address.
+     * @dev `msg.sender` must have received push access through `mate()`.
+     */
+    function quit() external isMate {
+        _doQuit(gem.balanceOf(address(this)));
+    }
+
+    /**
+     * @notice Flushes out specific amount of GEM balance to `quitTo` address.
+     * @dev `msg.sender` must have received push access through `mate()`.
+     * @param amt Gem amount.
+     */
+    function quit(uint256 amt) external isMate {
+        _doQuit(amt);
+    }
+
+    /**
+     * @notice Flushes out all outstanding DAI balance to `usr` address.
+     * @dev Can only be called by the admin
+     * @param usr Destination address.
+     */
+    function yank(address usr) external auth {
+        uint256 wad = dai.balanceOf(address(this));
+        dai.transfer(usr, wad);
+        emit Yank(usr, wad);
+    }
+
+    /**
+     * @notice Calculate required amount of GEM to get `wad` amount of DAI.
+     * @param wad DAI amount.
+     * @return gemAmt Amount of GEM required.
+     */
+    function requiredGemAmt(uint256 wad) external view returns (uint256 gemAmt) {
+        return mul(wad, WAD) / mul(sub(WAD, psm.tin()), to18ConvertionFactor);
+    }
+
+    /**
+     * @notice Swaps the specified amount of GEM into DAI through the PSM and push it into the `to` address.
+     * @param amt GEM amount.
+     */
+    function _doPush(uint256 amt) internal {
+        uint256 prevDaiBalance = dai.balanceOf(address(this));
+
+        psm.sellGem(address(this), amt);
+
+        uint256 daiPushAmt = sub(dai.balanceOf(address(this)), prevDaiBalance);
```

Comment:
If `daiPushAmt` is calculated locally, you can do the same than I proposed for the output conduit and `sellGem` directly to `to` address.

Reply:
But here we are not calculating it, to be able to do that we need to add an additional method that will be used only for this calculation. Not sure if we really need that, now we have 2 calls for `balanceOf` instead of having 1 call for `psm.tin()`???

Reply:
I'm talking about saving store operation, if you calculate `daiPushAmt` locally (to log in the event), you can do `sellGem` directly to the `to` address.

Reply:
Done